runtime/lua/vim/_extui/messages.lua

```diff
@@ -0,0 +1,394 @@
+local api, fn, o = vim.api, vim.fn, vim.o
+local ext = require('vim._extui.shared')
+
+---@class vim._extui.messages
+local M = {
+  -- Message box window. Used for regular messages with 'cmdheight' == 0 or,
+  -- cfg.msg.pos == 'box'. Also used for verbose messages regardless of
+  -- cfg.msg.pos. Automatically resizes to the text dimensions up to a point,
+  -- at which point only the most recent messages will fit and be shown.
+  -- A timer is started for each message whose callback will remove the message
+  -- from the window again.
+  box = {
+    count = 0, -- Number of messages currently in the message window.
+    width = 1, -- Current width of the message window.
+    timer = nil, ---@type uv.uv_timer_t Timer that removes the most recent message.
+  },
+  -- Cmdline message window. Used for regular messages with 'cmdheight' > 0.
+  -- Also contains 'ruler', 'showcmd' and search_cmd/count messages as virt_text.
+  -- Messages that don't fit the 'cmdheight' are cut off and virt_text is added
+  -- to indicate the number of spilled lines and repeated messages.
+  cmd = {
+    count = 0, -- Number of messages currently in the message window.
+    lines = 0, -- Number of lines in cmdline buffer (including wrapped lines).
+    msg_row = -1, -- Last row of message to distinguish for placing virt_text.
+    last_col = o.columns, -- Crop text to start column of 'last' virt_text.
+  },
+  dupe = 0, -- Number of times message is repeated.
+  prev_msg = '', -- Concatenated content of the previous message.
+  virt = { -- Stored virt_text state.
+    last = { {}, {}, {}, {} }, ---@type MsgContent[] status in last cmdline row.
+    msg = { {}, {} }, ---@type MsgContent[] [(x)] indicators in message window.
+    msg_hl = 0, ---@type integer? indicator hl_group, matching message tail.
+    idx = { mode = 1, search = 2, cmd = 3, ruler = 4, spill = 1, dupe = 2 },
+    ids = {}, ---@type { ['last'|'msg']: integer? } Table of mark IDs.
+  },
+}
+
+--- Start a timer whose callback will remove the message from the message window.
+---
+---@param buf integer Buffer the message was written to.
+---@param len integer Number of rows that should be removed.
+function M.box:start_timer(buf, len)
+  self.timer = vim.defer_fn(function()
+    if self.count == 0 or not api.nvim_buf_is_valid(buf) then
+      return -- Messages moved to split or buffer was closed.
+    end
+    api.nvim_buf_set_lines(buf, 0, len, false, {})
+    self.count = self.count - 1
+    -- Resize or hide message box for removed message.
+    if self.count > 0 then
+      M.set_pos('box')
+    else
+      self.width = 1
+      M.prev_msg = ext.cfg.msg.pos == 'box' and '' or M.prev_msg
+      api.nvim_win_set_config(ext.wins[ext.tab].box, { hide = true })
```

Comment:
Need to check `nvim_win_is_valid`: I have `vim.keymap.set('n', '<Esc>', vim.cmd.fclose)` and sometimes I hide the window before its timer expires.